### PR TITLE
Python provider QEP 122

### PR DESCRIPTION
**Pull request by elpaso**
## Python vector data providers

See: https://github.com/qgis/QGIS-Enhancement-Proposals/issues/122

qgis-python-provider

## Notes

1. I'm not in a hurry, we can wait for 3.4, but since the core changes are really limited I'd probably merge it now.
2. there is still a failing test because the provider names are hardcoded in `QgsVectorLayer::encodedSource`, but this is not a blocker for using the python provider
3. I really need a review for the SIP method code, I'm 100% not sure about what I did and memory leaks

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by elpaso** [comment]
@nyalldawson I think it's a good idea and whatever help modernizing the providers code is more than welcome.

**Comment by nyalldawson** [comment]
@elpaso I've opened #7035 adding this option

**Comment by elpaso** [comment]
@nyalldawson I'm not sure about the SIP bindings, would you please have a look?
https://github.com/qgis/QGIS/pull/7012/files#diff-5d57c8d6d109aeacb29ed3ba82e13313R67

File: src/core/qgsprovidermetadata.h [hunk 1]
```diff
@@ -58,10 +58,49 @@ class CORE_EXPORT QgsProviderMetadata
...
+      if ( sipResObj )
+      {
+        if ( sipCanConvertToType( sipResObj, sipType_QgsDataProvider, SIP_NOT_NONE ) )
+        {
+          int state0;
+          int sipIsErr = 0;
+          provider = reinterpret_cast<QgsDataProvider *>( sipConvertToType( sipResObj, sipType_QgsDataProvider, NULL, SIP_NOT_NONE, &state0, &sipIsErr ) );
+          if ( sipIsErr != 0 )
+          {
+            provider = nullptr;
```
Comment (nyalldawson) [inline src/core/qgsprovidermetadata.h]: I think you need to sipReleaseType( provider, ... ) here.

File: src/core/qgsproviderregistry.cpp [hunk 2]
```diff
@@ -525,6 +525,12 @@ void QgsProviderRegistry::registerGuis( QWidget *parent )
   }
 }
 
+void QgsProviderRegistry::registerProvider( QgsProviderMetadata *providerMetadata )
+{
+  if ( providerMetadata )
```
Comment (nyalldawson) [inline src/core/qgsproviderregistry.cpp]: Better check for existing providers with the same key here, and if so fail the registration. You could make this return a bool to indicate failures.

File: tests/src/python/test_provider_python.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+        pass
+
+    def testGetFeaturesNoGeometry(self):
+        """ Override and skip this test for pythonprovider provider, as it's actually more efficient for the pythonprovider provider to return
+        its features as direct copies (due to implicit sharing of QgsFeature)
+        """
+        pass
+
+    @unittest.skipIf(os.environ.get("TRAVIS", False), "Skipped on Travis")
+    def testGetFeaturesDestinationCrs(self):
```
Comment (nyalldawson) [inline tests/src/python/test_provider_python.py]: Is this still the case? where does it fail?

File: src/core/qgsmaplayer.cpp [hunk 4]
```diff
@@ -901,6 +901,7 @@ QString QgsMapLayer::loadNamedProperty( const QString &uri, QgsMapLayer::Propert
 {
   QgsDebugMsgLevel( QString( "uri = %1 myURI = %2" ).arg( uri, publicSource() ), 4 );
 
+  QgsDebugMsgLevel( "loadNamedProperty", 2 );
```
Comment (m-kuhn) [inline src/core/qgsmaplayer.cpp]: Is it so important?

**Comment by m-kuhn** [comment]
Nice to see this coming, it's been a long time to make this come true.

File: src/core/qgsmaplayer.cpp (hunk 4 above)
Comment (elpaso) [reply src/core/qgsmaplayer.cpp]: umm, not at all, it must be a leftover. I'll remove it.

File: src/core/qgsmaplayer.cpp [hunk 5]
```diff
@@ -912,7 +912,7 @@ QString QgsMapLayer::loadNamedProperty( const QString &uri, QgsMapLayer::Propert
   QFile myFile( uri );
   if ( myFile.open( QFile::ReadOnly ) )
   {
-    QgsDebugMsg( QString( "file found %1" ).arg( uri ) );
+    QgsDebugMsgLevel( QString( "file found %1" ).arg( uri ), 2 );
```
Comment (m-kuhn) [inline src/core/qgsmaplayer.cpp]: Is that one important enough ;) ?

File: tests/src/python/provider_python.py [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+
+    def __init__(self, source, request):
+        super().__init__(request)
+        self._request = request if request is not None else QgsFeatureRequest()
+        self._source = source
+        self._index = 0
+        self._transform = QgsCoordinateTransform()
+        if self._request.destinationCrs().isValid() and self._request.destinationCrs() != self._source._provider.crs():
+            self._transform = QgsCoordinateTransform(self._source._provider.crs(), self._request.destinationCrs(), self._request.transformContext())
+        self._filter_rect = self.filterRectToSourceCrs(self._transform)
```
Comment (nyalldawson) [inline tests/src/python/provider_python.py]: Here's why the test is failing on Travis- you need to wrap this in a try block:


    try:
        self._filter_rect = self.filterRectToSourceCrs(self._transform)
    except QgsCsException as e:
        self.close()
        return

Comment (nyalldawson) [reply tests/src/python/provider_python.py]: Yes - it's very tricky to get the cs exceptions thrown consistently. They seem to be very platform dependent.

File: src/core/qgsmaplayer.cpp (hunk 5 above)
Comment (elpaso) [reply src/core/qgsmaplayer.cpp]: what do you mean? make it a 4 ?

**Pull request merged**